## ЧАСТЬ I. Решите задания используя только стандартные методы Python 3

### 1. Импортируйте число e из библиотеки math и выведите его с окргулением до 4 цифр после запятой

In [97]:
import numpy as np
from math import e
from math import sqrt


print(round(e, 4))

2.7183


### 2. Вычислите целую часть результата деления 100 на 7. Убедитесь, что её тип – целое число.

In [16]:
result = 100 // 7

print(result)

if (type(result) == int):
    print('result is integer')

14
result is integer


### 3. Создайте список с квадратами целых чисел от 12 до 4 с шагом 2 не делящимися на 5.

In [35]:
squares = [i**2 for i in reversed(range(4, 13, 2)) if i**2 % 5 != 0]
print(squares)

[144, 64, 36, 16]


### 4.	Отсортируйте полученный в пункте 3 список относительно модуля разницы между элементами этого списка и числом 50.

In [41]:
squares.sort(key = lambda x: abs(x - 50))
print(squares)

[64, 36, 16, 144]


### 5. Создайте множество (set) содержащее все 26 маленькие буквы латинского алфавита.

In [47]:
alphabet = set(map(chr, range(ord('a'), ord('z') + 1))) # all lowercase letters from a to z
print(alphabet)
print(len(alphabet))

{'k', 'a', 'v', 'r', 'l', 'q', 'm', 'g', 'x', 'c', 't', 'd', 'y', 'p', 'b', 'n', 'z', 'i', 'h', 'j', 'u', 'w', 'f', 's', 'e', 'o'}
26


### 6.	Создайте список, содержащий числа с повторениями. Используя словарь, подсчитайте, сколько раз каждое число встречалось в этом списке, и напечатайте содержимое словаря

In [54]:
numbers = [i % 15 for i in range(100)]

occurrences = {}

for key in numbers:
    occurrences[key] = occurrences.get(key, 0) + 1

print(occurrences)

{0: 7, 1: 7, 2: 7, 3: 7, 4: 7, 5: 7, 6: 7, 7: 7, 8: 7, 9: 7, 10: 6, 11: 6, 12: 6, 13: 6, 14: 6}


### 7. Реализуйте функцию от трёх аргументов a, b и c находящую вещественные корни уравнения ax2 + bx + c = 0, с нулевыми значениями аргументов по умолчанию

In [87]:
def get_roots(a = 0, b = 0, c = 0):
    discriminant = b**2 - 4*a*c
    x = None
    y = None

    if (discriminant >= 0):
        x1 = (-b + sqrt(discriminant)) / (2*a)
        x2 = (-b - sqrt(discriminant)) / (2*a)

    return (x1, x2)

print(get_roots(3, -14, -5))

(5.0, -0.3333333333333333)


### 8.	Реализуйте класс Point с полями x,y заполняемыми в конструкторе

In [88]:
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def calc_euclid_dist(self, point: Point):
        return sqrt((self.x - point.x)**2 + (self.y - point.y)**2)


point_inside = Point(4, 10)
point_outside = Point(5, 12)

### 9. Реализуйте класс Circle наследника класса Point с дополнительным полем r и методом inside(p) проверяющим, что круг одержит точку p.

In [89]:
class Circle(Point):
    def __init__(self, x, y, radius):
        super().__init__(x, y)
        self.r = radius

    def inside(self, p: Point):
        return ((p.x - self.x)**2 + (p.y - self.y)**2) <= self.r**2


circle = Circle(1, 2, 9)
print(circle.inside(point_inside))
print(circle.inside(point_outside))

True
False


### 10. Реализуйте метод класса Point, принимающий другой объект класса Point и вычисляющий эвклидово расстояние до него

In [90]:
print(point_outside.calc_euclid_dist(point_inside))

2.23606797749979


## ЧАСТЬ II. Решите задачи используя библиотеку NumPy

### 11. Создайте матрицу 4x4, значения элементов которой варьируются от 0 до 15

In [110]:
m44 = np.arange(16).reshape(4, 4)
print(m44)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


### 12. Найдите индексы ненулевых элементов из списка [1,2,0,0,4,0,5,7,8,0,0,0]

In [125]:
lst = np.array([1, 2, 0, 0, 4, 0, 5, 7, 8, 0, 0, 0])

indices = np.where(lst != 0)
print(indices)

(array([0, 1, 4, 6, 7, 8]),)


### 13. Создайте единичную матрицу размера 3х3

In [112]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

### 14. Создайте матрицу 10х10, заполните её случайными значениями и найдите максимальный и минимальный элементы

In [113]:
m1010 = np.random.random((10, 10))
m1010_min = m1010.min()
m1010_max = m1010.max()

print(m1010)
print(m1010_min, m1010_max)

[[0.97570754 0.9415603  0.68955242 0.97822176 0.48589297 0.92609073
  0.82278342 0.29607358 0.46633526 0.64289143]
 [0.59952588 0.99908089 0.11401495 0.72290111 0.26720927 0.72621967
  0.40668188 0.78775461 0.164192   0.07384016]
 [0.78333242 0.1551884  0.49952477 0.27549214 0.64681323 0.65266275
  0.23203103 0.50375742 0.08226258 0.27342841]
 [0.23963302 0.91874146 0.05805325 0.1581931  0.71655229 0.18805965
  0.18917291 0.44786869 0.33009786 0.07796172]
 [0.1234716  0.45602223 0.12850014 0.91491613 0.58158392 0.13147724
  0.78055092 0.96305136 0.30045586 0.26201283]
 [0.40959409 0.07774513 0.1672518  0.55190554 0.5213492  0.784464
  0.97200928 0.58074422 0.15389354 0.1808573 ]
 [0.0933268  0.43851815 0.87630674 0.91099068 0.3853804  0.7338818
  0.0245075  0.70625451 0.46314141 0.40345297]
 [0.99376708 0.6402088  0.31391382 0.33508592 0.13780541 0.47675466
  0.43655936 0.86398245 0.05811647 0.19208266]
 [0.47644614 0.63588292 0.23906631 0.11424504 0.48673916 0.71601184
  0.12888444 0.

### 15. Создайте матрицу 5х5 и расположите в ней числа 1, 2, 3, 4 под главной диагональю

In [114]:
m55 = np.diag(np.arange(1, 5), k = -1)
print(m55)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


### 16. Перемножьте случайные матрицы размерностей 5х3 и 3х2

In [115]:
m53 = np.random.random((5, 3))
m32 = np.random.random((3, 2))

print(np.dot(m53, m32))

[[0.17602783 0.71928536]
 [0.45948143 0.65725988]
 [0.35434624 0.70015631]
 [0.42901136 0.87656213]
 [0.51077894 0.68492945]]


### 17. Найдите число из списка t, наиболее близкое по величине к числу a

In [142]:
a = 2.39
t = [0, 1, 1.2, 2, 3, 9]

new_t = np.abs(np.array(t) - a)
index = new_t.argmin()
result = t[index]

print(result)

2


### 18. Создайте массив из 10 элементов и поменяйте значение его наибольшего элемента на 0 

In [143]:
m_max = np.random.random((10))
m_max[m_max.argmax()] = 0
print(m_max)

[0.47570801 0.65469895 0.60431323 0.18284506 0.32658258 0.27363464
 0.40504362 0.         0.31452497 0.60485759]


### 19. Создайте матрицу 5x5 и поменяйте в ней местами 2 и 3 строку

In [151]:
matrix = np.random.randint(0, 9, size = (5, 5))
print(matrix)
print('\n')

matrix[2], matrix[3] = matrix[3], matrix[2].copy()
print(matrix)

[[5 7 0 7 4]
 [0 6 0 8 8]
 [1 8 6 1 0]
 [1 5 8 8 0]
 [8 7 5 0 2]]


[[5 7 0 7 4]
 [0 6 0 8 8]
 [1 5 8 8 0]
 [1 8 6 1 0]
 [8 7 5 0 2]]


### 20. Создайте случайную целочисленную матрицу и найдите наиболее часто встречающийся в ней элемент

In [207]:
n = np.random.randint(2, 20)
m = np.random.randint(2, 20)

matrix = np.random.randint(0, 100, size = (n, m))
print(matrix)

unique = np.unique(matrix, return_counts=True)
index = unique[1].argmax()
result = unique[0][index]

print(result)

[[ 0 55 78 60 68 80]
 [61 74 30 34 99 66]
 [34 37 92  2 15 55]
 [ 0  8 86 89 83 34]
 [74 71 67  5 77  5]
 [73 55  5 43 33 92]
 [89 95 20 36 66 85]
 [70 36 68 56 29 39]
 [77 56 40  5 53 95]
 [22 50 74 11 17  4]
 [ 3 72 30 54 12 74]
 [80 26 96 58 45 71]
 [84 88 58 28 55 34]
 [ 1  8 67 32 76  3]
 [55 59 68  5  0 94]
 [47 63 84 19 63 24]
 [ 1 90 92 29 36 43]
 [56 58 43 41 26 72]]
5


## Часть III. Прочитайте краткий мануал по библиотеке pandas и выполните оставшиеся задания

Pandas - это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов. В связке с библиотеками Matplotlib и Seaborn появляется возможность удобного визуального анализа табличных данных.


Для использования библиотек их необходимо импортировать, что можно сделать с помощью команды import. После ключевого слова import указывается название библиотеки, а после as - ее краткое название (алиас), которое мы будем использовать в коде для краткого обозначения этой библиотеки.

In [0]:
import pandas as pd

Основными структурами данных в Pandas являются классы Series и DataFrame. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа (его можно интеерпретировать как индексированный столбец таблицы). Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа Series. Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

## Чтение из файла и первичный анализ
Чаще всего, данные, которые мы будем анализировать, готовить  и использовать для обучения моделей, не формируются самостоятельно, а берутся из готового дататсета. Рассмотрим чтение датасета (DataFrame) из файла с расширением csv. В качестве примера возьмем titanic_train.csv, который содержить данные о пассажирах Титаника. Полный датасет titanic был разделен на 2 части - тренировочную и тестовую. Именно с тренировочной частью мы и будем работать. Датасет titanic предоставлен в рабочей папке Лаб_1, а также вы можете его скачать (titanic_train.csv) по ссылке https://www.kaggle.com/tedllh/titanic-train#titanic_train.csv. Если вы работаете на платформе Goggle Colab, необходимо сделать upload файла в вашу рабочую директорию ( в меню ноутбука слева  найти и перейти на Files, нажать Upload и выбрать файл для загрузки)

Внимание! Проверьте, что файл находиться в папке content, если его там нет, то нажмите на нее правой кнопкой и upload. 


### Чтение данных

Для чтения csv файла используем функцию pd.read_csv

In [152]:
import pandas as pd
df = pd.read_csv('titanic_train.csv')

Посмотрим тип переменной df

In [153]:
type(df)

pandas.core.frame.DataFrame

### Первичный анализ

После того, как мы считали данные, посмотрим первые несколько записей датафрейма (если в скобках после функции head не указать нужное количество записей, то по умолчанию будут выведены первые 5 записей датафрейма). Если вместо таблицы вы видите  "кашу" из данных, разделенных ";", значит, ваш csv файл использует в качестве разделителя ";", а не ",". И для корректного чтения данных, нужно задать разделитель в функции чтения, например df = pd.read_csv('proba.csv', sep=";")

In [154]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Давайте посмотрим, что это за данные. Достаточно часто датасеты имеют краткое описание содержимого. Например, что означает тот или иной столбец, какие значения используются для заполнения и т.д. Или значения признаков интуитивно понятны. В нашем случае, мы имеем следующие данные:
каждая строка - это сведения об отдельном пассажире. Строка содержит:
*   PassengerId - идентификатор пассажира (проще говоря, порядковый номер строки)
*   Survived - признак, выжил ли пассажир после крушения (0 - погиб, 1 - выжил)
*   Pclass - в каком классе ехал пассажир (1, 2 или 3)
*   Name - имя пассажира
*   Sex - пол (male - мужской, female - женский)
*   Age - возраст
*   SibSp - количество родственников (по горизонтали - супруг, братья, сестры) на борту
*   Parch - количество родственников (по вертикали - родители, дети) на борту 
*   Ticket - номер билета
*   Fare - цена билета
*   Cabin - номер каюты
*   Embarked - порт посадки (C — Cherbourg, Q — Queenstown, S — Southampton)





Команда shape() показывате размерность датафрейма

Мы видим, что в нашем датафрейме 891 строка (то есть, сведения о 891 пассажире) и 12 столбцов признаков

In [155]:
df.shape

(891, 12)

Команда columns.values() показывате название всех признаков

In [156]:
df.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

Также полезно посмотреть метод info, который показывает краткую статистику по набору данных

In [157]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Можно увидеть, что не все признаки у нас заполнены в каждой строке,для некоторых признаков есть пропуски. Например, у нас 891 строка, а Age (возраст) заполнен только в 714 строках. Совсем мало значений в столбце Cabin. 

Есть несколько вариантов работы с такими строками: 
* их можно оставить без изменений, если признак, в котором пропущены данные, малозначим;
* можно удалить все строки, в которых есть пустые значения;
* можно заполнить их какими-либо значениями (0, медианой или любыми другими константами).

Метод info так же показывает какого типа данные в том или ином столбце: цифровые (int или float), текстовые и т.д. В последней строке указан объем памяти, которую занимает датафрейм. При больших объемах данных это позволяет оценить хватит ли вам ресурсов для загрузки в память полного датасета

Еще один полезный метод для начального анализа данных - метод describe(), который выдает краткую статистику по числовым признакам: среднее, стандартное отклонение, минимум, максимум, медианы, квартили и т.д. Самостоятельно посмотрите, что означает каждое понятие. 

In [158]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Доступ к элементам данных

### Доступ к элементам DataFrame

Разберемся как посмотреть какие-то конкретные значения в датафрейме. Наш датафрейм представляет собой таблицу, состоящую из строк и столбцов. Обратиться к конкретному столбцу можно используя его имя. Например, выведем первые 10 строк из столбца с возрастом пассажиров

In [159]:
df['Age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Второй способ обращения - через "."

In [160]:
df.Age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Столбец таблицы - это объек Pandas Series, убедимся в этом определив тип полученного объекта. Этот объект проиндексирован (слева печатаются индексы). В нижней строке есть информацию о его имени и типе значений.
Кроме того, здесь хорошо видны пропуски (незаполненные значения) в некоторых строках (они обозначены NaN - Not-A-Number)

In [161]:
type(df['Age'])

pandas.core.series.Series

Кроме обращения к столбцу по имени, можно использовать индексацию с помощью методов loc и iloc

Метод loc выдает нам подматрицу, используя индексы строк и столбцов. Первый аргумент отвечает за количество записей (как с начала, так и при желании с любого места), а второй - за отраженные в подматрице признаки.


In [162]:
df.loc[0:5, ['Pclass', 'Fare']]

Pclass     Fare
0       3   7.2500
1       1  71.2833
2       3   7.9250
3       1  53.1000
4       3   8.0500
5       3   8.4583

Метод iloc работает схожим образом, за исключением того, что второй аргумент теперь тоже индексы, то есть отражает количество признаков.


In [0]:
df.iloc[0:5, 1:3]

# Доступ к элементам Series

Для примера создадим два объекта Series

In [163]:
s1 = pd.Series(["one", "two", "three", "four", "five"]) # передаем список, индексы проставляются автоматически с 0
print(s1)

0      one
1      two
2    three
3     four
4     five
dtype: object


In [164]:
s2 = pd.Series([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e'])  # передаем список, индексы заданы символами от 'a' до 'e'
print(s2)

a    1
b    2
c    3
d    4
e    5
dtype: int64


К элементам Series можно обращаться по численному индексу через "[ ]"

In [165]:
s2[2] # второй элемент серии s2 (индексация с 0)

3

К элементам Series можно обращаться по метке:

In [166]:
s2['c']

3

К элементам Series можно применять так называемые slice (срезы) через ':'.
Срезами является часть серии до или после определенного элемента

In [0]:
s2[:2] # все элементы до второго включительно

In [0]:
s2[2:] # второй и все элементы после второго до конца

In [0]:
s2[2:4] # все элементы со второго по четвертый не включительно. Последний никогда не включается.

In [0]:
s2[:] # все элементы

В поле для индекса можно поместить условное выражение.
Данное выражение можно перевести как вопрос, является ли значение элемента (признака) меньшим или равным двум.

In [0]:
s2[s2 <= 2]

Или даже несколько условных выражений, отделенных "( )". 

Если соединить два условия знаком "&", то результат будет удовлетворять двум условиям, а если знаком "|", то хотя бы одному из них.

In [0]:
s2[(s2 >= 2) & (s2 <=4)]

In [0]:
s2[(s2 <= 2) | (s2 == 5)]

Важно, что в условных выражениях равенство пишется как "==" (два знака "равно").

Со структурами Series можно работать как с векторами: складывать, умножать вектор на число и т.п.

In [0]:
s3 = s2 + 10

In [0]:
print(s3)

## Примеры анализа информации

Для примера отберем пассажиров мужского пола, возраст которых меньше сорока лет.

In [0]:
df[(df['Sex'] == 'male') & (df.Age < 40)].head(5)

Можно отсортировать этих людей по возрастанию платы за билет. Ascending в данной функции отвечает за сортировку по убыванию (False) или по возрастанию (True).

In [0]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).head(10)

Можно подсчитать количество пассажиров (а точнее их уникальных номеров) с помощью функции count()

In [0]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).PassengerId.count()

Также можно быстро вывести распределение переменной Sex с помощью функции value_count()

In [0]:
df.Sex.value_counts()

## Задания

### 21. Сколько человек спаслись и сколько погибли?

### 22. Выведите распределение переменной Survived отдельно для мужчин и отдельно для женщин


### 23. Каковы медиана и стандартное отклонение цен за билет?

### 24. Подсчитайте долю выживших среди людей моложе 25 лет и старше 55.

### 25. Подсчитайте долю выживших среди мужчин и женщин.

### 26. В каком городе больше всего село пассажиров?

### 27. Подсчитайте средний возраст выживших

### 28. Подсчитайте средний возраст пассажиров каждого из трех классов


### 29. Подсчитайте средний возраст выживших мужчин, купивших билет во 2 класс

### 30. Оцените, как влияет на выживаемость пассажиров наличие их родственников на борту.